[SPARK-54541][SQL] Rename _LEGACY_ERROR_TEMP_1007 and add sqlState

AbinayaJayaprakasam opened the pull request:
### What changes were proposed in this pull request?

I renamed _LEGACY_ERROR_TEMP_1007 to VIEW_WRITE_NOT_ALLOWED and added sqlState 42809.

The changes include:
- New error class VIEW_WRITE_NOT_ALLOWED in error-conditions.json with sqlState 42809
- Updated QueryCompilationErrors.scala to reference the new error class
- Refactored 6 test cases in DataFrameWriterV2Suite to use checkError() instead of string matching

### Why are the changes needed?

Users frequently encounter this error when attempting to write to temporary views using the DataFrameWriter V2 API. The legacy error name is non-descriptive and lacks an sqlState, which makes it difficult to handle programatically and reduces compatibility with JDBC/ODBC clients.

### Does this PR introduce any user-facing change?

No. The error message text stays the same. User will only see the same error message, but the underlying error class is now properly named and includes a sqlState for better tooling support.

### How was this patch tested?

I tested this change in multiple ways:
- Manually triggered the error in spark-shell to verify the new error class is used
- Updated 6 existing tests in DataFrameWriterV2Suite to use checkError() for structured validation of error class and parameters
- Ran the full test suite: all tests pass successfully


#### Manual Testing in Spark-Shell

**Before (legacy error):**


```
scala> val sourceData = Seq((1, "a"), (2, "b"), (3, "c")).toDF("id", "name")
scala> sourceData.createOrReplaceTempView("source")
scala> spark.range(10).createOrReplaceTempView("temp_view")
scala> spark.table("source").writeTo("temp_view").append()

org.apache.spark.sql.AnalysisException: Cannot write into temp view temp_view as it's not a data source v2 relation.
at org.apache.spark.sql.errors.QueryCompilationErrors$.writeIntoTempViewNotAllowedError(QueryCompilationErrors.scala:551)
at org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveRelations$$anonfun$apply$12.$anonfun$applyOrElse$73(Analyzer.scala:1237)

```


**After (new error with class and sqlState):**
```
scala> val sourceData = Seq((1, "a"), (2, "b"), (3, "c")).toDF("id", "name")
scala> sourceData.createOrReplaceTempView("source")
scala> spark.range(10).createOrReplaceTempView("temp_view")
scala> spark.table("source").writeTo("temp_view").append()

org.apache.spark.sql.AnalysisException: [VIEW_WRITE_NOT_ALLOWED] Cannot write into view temp_view, please write into a table instead. SQLSTATE: 42809
  at org.apache.spark.sql.errors.QueryCompilationErrors$.writeIntoTempViewNotAllowedError(QueryCompilationErrors.scala:551)
  at org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveRelations$$anonfun$apply$12.$anonfun$applyOrElse$73(Analyzer.scala:1237)
```
  

*Note: The error now displays the error class `[VIEW_WRITE_NOT_ALLOWED]` and includes sqlState `42809` (accessible via `getSqlState()`).*

Ran the full DataFrameWriterV2Suite test suite: all tests pass successfully

```
[info] Test run finished: 0 failed, 0 ignored, 6 total, 0.602s
[info] ScalaTest
[info] Run completed in 8 seconds, 683 milliseconds.
[info] Total number of tests run: 48
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 48, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Passed: Total 54, Failed 0, Errors 0, Passed 54
[success] Total time: 38 s, completed 13-Dec-2025, 10:51:11 pm
```


### Was this patch authored or co-authored using generative AI tooling? 
No



<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

AbinayaJayaprakasam committed:
[SPARK-54541][SQL] Rename _LEGACY_ERROR_TEMP_1007 and add sqlState

### What changes were proposed in this pull request?

I renamed _LEGACY_ERROR_TEMP_1007 to TEMP_VIEW_WRITE_NOT_ALLOWED and added sqlState 42809.

The changes include:
- New error class TEMP_VIEW_WRITE_NOT_ALLOWED in error-conditions.json with sqlState 42809
- Updated QueryCompilationErrors.scala to reference the new error class
- Refactored 3 test cases in DataFrameWriterV2Suite to use checkError() instead of string matching
- Removed the old _LEGACY_ERROR_TEMP_1007 entry to complete the migration

### Why are the changes needed?

Users frequently encounter this error when attempting to write to temporary views using the DataFrameWriter V2 API. The legacy error name is non-descriptive and lacks a sqlState, which makes it difficult to handle programmatically and reduces compatibility with JDBC/ODBC clients.

### Does this PR introduce any user-facing change?

No. The error message text stays the same. Users will see the same error message, but the underlying error class is now properly named and includes a sqlState for better tooling support.

### How was this patch tested?

I tested this change in multiple ways:
- Manually triggered the error in spark-shell to verify the new error class is used
- Updated 3 existing tests in DataFrameWriterV2Suite to use checkError() for structured validation of error class and parameters
- Ran the full test suite: all 48 tests pass successfully
- Verified sqlState 42809 is correctly set based on reviewer feedback

### Was this patch authored or co-authored using generative AI tooling?

no

AbinayaJayaprakasam commented on common/utils/src/main/resources/error/error-conditions.json:
```diff
@@ -6029,6 +6029,12 @@
+    "message" : [
+      "Cannot write into temp view <quoted> as it's not a data source v2 relation."
+    ],
+    "sqlState" : "42623"
```
Thank you .. You're absolutely right. 

I initially chose 42623 because I saw it used for other write constraint errors in Spark, but after reviewing the SQL state definitions, 42809 ("The identified object is not the type of object to which the statement applies") is indeed more semantically accurate for this error.

The error occurs because the temp view is not a DataSource V2 relation, which is exactly an "object type mismatch" scenario that 42809 is designed for.

I've updated the PR to use 42809. Thank you for the careful review!

**References:**
- IBM DB2: [42809 - Wrong object type](https://www.ibm.com/docs/en/db2/11.5?topic=messages-sqlstate)
- PostgreSQL: [42809 - Wrong object type](https://www.postgresql.org/docs/current/errcodes-appendix.html)

cloud-fan commented on common/utils/src/main/resources/error/error-conditions.json:
```diff
@@ -6029,6 +6029,12 @@
   },
+  "TEMP_VIEW_WRITE_NOT_ALLOWED" : {
+    "message" : [
+      "Cannot write into temp view <quoted> as it's not a data source v2 relation."
```
I found the v1 error: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L1220

Can we reuse it?

AbinayaJayaprakasam committed:
[SPARK-54541][SQL] Unify view write errors into VIEW_WRITE_NOT_ALLOWED

This commit unifies the error handling for writing to views (both temp and persistent) in the V2 API:

- Renamed TEMP_VIEW_WRITE_NOT_ALLOWED to VIEW_WRITE_NOT_ALLOWED for broader coverage
- Removed legacy error _LEGACY_ERROR_TEMP_1011
- Updated error message to work for both temp and persistent views
- Added sqlState 42809 (Wrong object type)
- Updated QueryCompilationErrors.scala to use the unified error in both methods
- Refactored 6 test cases to use checkError() for structured validation

The unified error provides better consistency and simplifies maintenance.

AbinayaJayaprakasam committed:
Fix alphabetical ordering of VIEW_WRITE_NOT_ALLOWED in error-conditions.json

Move VIEW_WRITE_NOT_ALLOWED to correct alphabetical position between
VIEW_NOT_FOUND and WINDOW_FUNCTION_AND_FRAME_MISMATCH to fix
SparkThrowableSuite test failure.

AbinayaJayaprakasam committed:
Address reviewer feedback: Improve error message and parameter naming

Updated VIEW_WRITE_NOT_ALLOWED error based on cloud-fan's suggestion:
- Changed error message to: 'Cannot write into view <name>, please write into a table instead.'
- Renamed parameter from 'quoted' to 'name' for consistency
- Updated all 6 test assertions to use the new parameter name

This provides a cleaner, more user-friendly error message that aligns
with the existing error pattern in Analyzer.scala.

cloud-fan commented:
the failed JDBC test is unrelated, thanks, merging to master!
